src/AspNet.Security.OAuth.Xiaomi/XiaomiAuthenticationHandler.cs

```diff
@@ -0,0 +1,166 @@
+        if (!response.IsSuccessStatusCode)
+        {
+            await Log.UserProfileErrorAsync(Logger, response, Context.RequestAborted);
+            throw new HttpRequestException("An error occurred while retrieving user information.");
+        }
+
+        var json = await response.Content.ReadAsStringAsync(Context.RequestAborted);
+        using var payload = JsonDocument.Parse(json);
+        var rootElement = payload.RootElement;
+
+        if (!rootElement.TryGetProperty("data", out JsonElement dataElement))
+        {
+            var errorCode = rootElement.GetString("code")!;
+            throw new Exception($"An error (Code:{errorCode}) occurred while retrieving user information.");
+        }
+
+        identity.AddClaim(new Claim(ClaimTypes.NameIdentifier, dataElement.GetString("unionId")!, ClaimValueTypes.String, Options.ClaimsIssuer));
+
+        var principal = new ClaimsPrincipal(identity);
+        var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options, Backchannel, tokens, dataElement);
+        context.RunClaimActions();
+
+        await Events.CreatingTicket(context);
+        return new AuthenticationTicket(context.Principal!, context.Properties, Scheme.Name);
+    }
+
+    protected override async Task<OAuthTokenResponse> ExchangeCodeAsync([NotNull] OAuthCodeExchangeContext context)
+    {
+        var tokenRequestParameters = new Dictionary<string, string?>()
+        {
+            ["client_id"] = Options.ClientId,
+            ["client_secret"] = Options.ClientSecret,
+            ["code"] = context.Code,
+            ["redirect_uri"] = context.RedirectUri,
+            ["grant_type"] = "authorization_code",
+        };
+
+        // PKCE https://tools.ietf.org/html/rfc7636#section-4.5, see BuildChallengeUrl
+        if (context.Properties.Items.TryGetValue(OAuthConstants.CodeVerifierKey, out var codeVerifier))
+        {
+            tokenRequestParameters.Add(OAuthConstants.CodeVerifierKey, codeVerifier!);
+            context.Properties.Items.Remove(OAuthConstants.CodeVerifierKey);
+        }
+
+        var address = QueryHelpers.AddQueryString(Options.TokenEndpoint, tokenRequestParameters);
+
+        using var response = await Backchannel.GetAsync(address, Context.RequestAborted);
+        if (!response.IsSuccessStatusCode)
+        {
+            await Log.ExchangeCodeErrorAsync(Logger, response, Context.RequestAborted);
+            return OAuthTokenResponse.Failed(new Exception("An error occurred while retrieving an access token."));
+        }
+
+        var json = await response.Content.ReadAsStringAsync(Context.RequestAborted);
+
+        // The json is a special string containing '&&&START&&&' and needs trim it.
+        // See https://dev.mi.com/console/doc/detail?pId=707#_0_1 for details.
+        if (json.StartsWith("&&&START&&&", StringComparison.OrdinalIgnoreCase))
+        {
+            json = json[11..];
+        }
```

Comment:
```suggestion
        // The JSON in the response has a prefix that needs to be removed.
        // See https://dev.mi.com/console/doc/detail?pId=707#_0_1 for details.
        const Prefix = "&&&START&&&";
        if (json.StartsWith(Prefix, StringComparison.Ordinal))
        {
            json = json[Prefix.Length..];
        }
```

Reply:
This is canonical, but the variable `Prefix` seems to be missing the type identifier.

Reply:
Well that's what happens when you try and write code in a browser and not an IDE 😉

Reply:
This is canonical, but the variable `Prefix` seems to be missing the type identifier.

Reply:
Yes, Just like my current network is terrible, anything can happen, I'll revise the commit again. :sweat_smile: